lib/pure/algorithm.nim

```diff
@@ -66,18 +66,32 @@ proc reversed*[T](a: openArray[T]): seq[T] =
 proc binarySearch*[T](a: openArray[T], key: T): int =
   ## binary search for `key` in `a`. Returns -1 if not found.
   var b = len(a)
-  while result < b:
-    var mid = (result + b) div 2
-    if a[mid] < key: result = mid + 1
-    else: b = mid
-  if result >= len(a) or a[result] != key: result = -1
-
-proc smartBinarySearch*[T](a: openArray[T], key: T): int =
+  if b == 0: return -1
+  if unlikely((b and (b - 1)) == 0):
```

Comment:
Why is this unlikely? I know a lot of data is constructed to have a power of two size which makes this less "unlikely". I would leave it as the default.